### PR TITLE
Implement ability to use default ingress TLS certificate

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

### What does this PR do
Adds ability to tell Che operator to use default ingress TLS certificate to secure Che endpoints.
To use default ingress certificate, field `k8s.tlsSecretName` should be empty string or absent.

### Which isseus this PR fixes
https://github.com/eclipse/che/issues/18079

### How to test

1. Create a new Minikube instance.
2. Enable ingress addon: `minikube addons enable ingress`
3. Change the default ingress certificate

<details>
<summary>Automation script</summary>

```sh
#!/bin/bash

MINIKUBE_DOMAIN=$(minikube ip).nip.io

CA_CN='Custom Minikube CA'
CA_KEY_FILE='ca.key'
CA_CERT_FILE='ca.crt'

CN_SERVER='Minikube Ingress'
SERVER_KEY_FILE='domain.key'
SERVER_CERT_REQUEST_FILE='domain.csr'
SERVER_CERT_FILE='domain.crt'

CHAIN_FILE='chain.pem'

# Detect openssl configuration file
OPENSSL_CNF='/etc/pki/tls/openssl.cnf'
if [ ! -f $OPENSSL_CNF ]; then
    OPENSSL_CNF='/etc/ssl/openssl.cnf'
fi

DIR=/tmp/minikube-cert
rm -rf $DIR && mkdir $DIR && cd $DIR

# Generate private key for root CA
openssl genrsa -out $CA_KEY_FILE 4096

# Generate root CA certificate and sign it with previously generated key.
openssl req -batch -new -x509 -nodes -key $CA_KEY_FILE -sha256 -subj /CN="${CA_CN}" -days 1024 -reqexts SAN -extensions SAN -config <(cat ${OPENSSL_CNF} <(printf '[SAN]\nbasicConstraints=critical, CA:TRUE\nkeyUsage=keyCertSign, cRLSign, digitalSignature')) -outform PEM -out $CA_CERT_FILE

# Generate server prvate key
openssl genrsa -out $SERVER_KEY_FILE 2048

# Create certificate request for the ingress
openssl req --batch -new -sha256 -key $SERVER_KEY_FILE -subj "/CN=${CN_SERVER}" -reqexts SAN -config <(cat $OPENSSL_CNF <(printf "\n[SAN]\nsubjectAltName=DNS:${MINIKUBE_DOMAIN},DNS:*.${MINIKUBE_DOMAIN}\nbasicConstraints=critical, CA:FALSE\nkeyUsage=digitalSignature, keyEncipherment, keyAgreement, dataEncipherment\nextendedKeyUsage=serverAuth")) -outform PEM -out $SERVER_CERT_REQUEST_FILE

# Create certificate for ingress domain based on given certificate request.
openssl x509 -req -in $SERVER_CERT_REQUEST_FILE -CA $CA_CERT_FILE -CAkey $CA_KEY_FILE -CAcreateserial -days 365 -sha256 -extfile <(printf "subjectAltName=DNS:${MINIKUBE_DOMAIN},DNS:*.${MINIKUBE_DOMAIN}\nbasicConstraints=critical, CA:FALSE\nkeyUsage=digitalSignature, keyEncipherment, keyAgreement, dataEncipherment\nextendedKeyUsage=serverAuth") -outform PEM -out $SERVER_CERT_FILE

# Create certificate chain file
cat $SERVER_CERT_FILE $CA_CERT_FILE > $CHAIN_FILE

# Create secret
SECRET_NAME='new-ingress-tls'
kubectl create secret tls ${SECRET_NAME} --key ${SERVER_KEY_FILE} --cert ${CHAIN_FILE} --namespace kube-system

# Patch nginix ingress controller
kubectl patch deployment ingress-nginx-controller --namespace kube-system --type=json -p="[{\"op\": \"add\", \"path\": \"/spec/template/spec/containers/0/args/-\", \"value\": \"--default-ssl-certificate=kube-system/${SECRET_NAME}\"}]"

echo "Minikube TLS certificate has been changed. New CA: ${DIR}/${CA_CERT_FILE}"

```

</details>

4. Deploy Eclipse Che with command: 
```
./run server:start --platform=minikube --installer=operator --che-operator-image=<image-built-from-this-PR> --che-operator-cr-patch-yaml default-secret-patch.yaml 
```

where `default-secret-patch.yaml` is:
```yaml
spec:
  k8s:
    tlsSecretName: ''
```
5. Check that everything works

### CODE REVIEW

File: deploy/olm-catalog/eclipse-che-preview-kubernetes/manifests/che-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -13,12 +13,12 @@ metadata:
           "spec": {
             "k8s": {
                 "ingressDomain": "",
-                "tlsSecretName": ""
+                "tlsSecretName": "che-tls"
```
Comment (AndrienkoAleksandr) [inline deploy/olm-catalog/eclipse-che-preview-kubernetes/manifests/che-operator.clusterserviceversion.yaml:16]: You provided this changes only for kuberentes csv file. We have openshift csv also.

File: deploy/crds/org_v1_che_crd.yaml [hunk 2]
```diff
@@ -271,7 +271,9 @@ spec:
                   type: string
                 tlsSecretName:
                   description: Name of a secret that will be used to setup ingress
-                    TLS termination if TLS is enabled. See also the `tlsSupport` field.
+                    TLS termination if TLS is enabled. If the field is empty string,
```
Comment (AndrienkoAleksandr) [inline deploy/crds/org_v1_che_crd.yaml:274]: Sorry, but where is changes in the che_types? I think you should change this text in the https://github.com/eclipse/che-operator/blob/master/pkg/apis/org/v1/che_types.go and then regenerated crd using https://github.com/eclipse/che-operator/blob/master/olm/update-crd-files.sh or you can use VSCode task https://github.com/eclipse/che-operator/blob/master/.vscode/tasks.json#L5

File: deploy/olm-catalog/eclipse-che-preview-kubernetes/manifests/che-operator.clusterserviceversion.yaml (hunk 1 above)
Comment (mmorhun) [reply deploy/olm-catalog/eclipse-che-preview-kubernetes/manifests/che-operator.clusterserviceversion.yaml:16]: This property is under `k8s` group.

Comment (tolusha) [reply deploy/olm-catalog/eclipse-che-preview-kubernetes/manifests/che-operator.clusterserviceversion.yaml:16]: It is not used in OpenShift

File: deploy/crds/org_v1_che_crd.yaml (hunk 2 above)
Comment (mmorhun) [reply deploy/crds/org_v1_che_crd.yaml:274]: Right, hank you. I'll make corresponding changes.

File: deploy/olm-catalog/eclipse-che-preview-kubernetes/manifests/che-operator.clusterserviceversion.yaml (hunk 1 above)
Comment (AndrienkoAleksandr) [reply deploy/olm-catalog/eclipse-che-preview-kubernetes/manifests/che-operator.clusterserviceversion.yaml:16]: Ok, sorry, but what about lines with set up devfileRegistryImage and devfileRegistryImage to empty value? https://github.com/eclipse/che-operator/pull/508/files#diff-d45591b58d736ef67fb4c34a130def83b8461a9f44998f0a3bae6d9aee239e66R20 for example

**Review by AndrienkoAleksandr — APPROVED** [review]

File: pkg/controller/che/che_controller.go [hunk 3]
```diff
@@ -437,21 +429,29 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
...
 		// Handle Che TLS certificates on Kubernetes infrastructure
 		if instance.Spec.Server.TlsSupport {
-			result, err := deploy.K8sHandleCheTLSSecrets(deployContext)
-			if result.Requeue || result.RequeueAfter > 0 {
-				if err != nil {
-					logrus.Error(err)
+			if instance.Spec.K8s.TlsSecretName != "" {
+				// Self-signed certificate should be created to secure Che ingresses
+				result, err := deploy.K8sHandleCheTLSSecrets(deployContext)
+				if result.Requeue || result.RequeueAfter > 0 {
```
Comment (tolusha) [inline pkg/controller/che/che_controller.go:443]: Simply.
```
If err {
 logrus.Error(...)
}
return result, err
```

File: pkg/deploy/ingress.go [hunk 4]
```diff
@@ -132,10 +136,8 @@ func getSpecIngress(
 		}
 	}
 
-	tls := "false"
-	tlsSecretName := util.GetValue(deployContext.CheCluster.Spec.K8s.TlsSecretName, "che-tls")
+	tlsSecretName := util.GetValue(deployContext.CheCluster.Spec.K8s.TlsSecretName, "")
```
Comment (tolusha) [inline pkg/deploy/ingress.go:139]:
```suggestion
	tlsSecretName :=  deployContext.CheCluster.Spec.K8s.TlsSecretName
```

**Review by flacatus — APPROVED** [review]

File: pkg/deploy/tls.go [hunk 5]
```diff
@@ -60,99 +61,141 @@ func IsSelfSignedCertificateUsed(deployContext *DeployContext) (bool, error) {
...
+				}
 
-		// Check the chain if ti contains self-signed CA certificate
-		for _, cert := range peerCertificates {
-			if cert.Subject.String() == cert.Issuer.String() {
-				// Self-signed CA certificate is found in the chain
+				// Both secrets (che-tls and self-signed-certificate) are missing which means that we should generate them (i.e. use self-signed certificate).
 				return true, nil
 			}
+			// TLS secret found, consider it as commonly trusted.
```
Comment (tolusha) [inline pkg/deploy/tls.go:80]: What about the case when self-signed-certificate secret exists but che-tls does not? How do we handle that ?

File: pkg/controller/che/che_controller.go (hunk 3 above)
Comment (mmorhun) [reply pkg/controller/che/che_controller.go:443]: No, it will cause infinite loop

File: pkg/deploy/tls.go (hunk 5 above)
Comment (mmorhun) [reply pkg/deploy/tls.go:80]: Operator will delete `self-signed-certificate` secret and will generate new pair.

**Review by tolusha — APPROVED** [review]

File: pkg/deploy/ingress.go (hunk 4 above)
Comment (mmorhun) [reply pkg/deploy/ingress.go:139]: Will fix later

**Pull request merged**